algorithms/src/Kokkos_Random.hpp

```diff
@@ -886,29 +886,27 @@ class Random_XorShift64_Pool {
  public:
   using generator_type = Random_XorShift64<DeviceType>;
 
-  KOKKOS_INLINE_FUNCTION
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+  KOKKOS_FUNCTION
   Random_XorShift64_Pool() {
-    num_states_ = 0;
-    padding_    = 0;
+    KOKKOS_IF_ON_DEVICE((
+        /* nonsensical initialization in the name of backward compatibility */
+        num_states_ = 0; padding_ = 0;))
+    KOKKOS_IF_ON_HOST(init(0, execution_space().concurrency());)
   }
+
   Random_XorShift64_Pool(uint64_t seed) {
     num_states_ = 0;
 
     init(seed, execution_space().concurrency());
   }
+#else
+  Random_XorShift64_Pool(uint64_t seed = 0) {
+    num_states_ = 0;
 
-  KOKKOS_INLINE_FUNCTION
-  Random_XorShift64_Pool(const Random_XorShift64_Pool& src)
-      : locks_(src.locks_), state_(src.state_), num_states_(src.num_states_) {}
-
-  KOKKOS_INLINE_FUNCTION
-  Random_XorShift64_Pool operator=(const Random_XorShift64_Pool& src) {
-    locks_      = src.locks_;
-    state_      = src.state_;
-    num_states_ = src.num_states_;
-    padding_    = src.padding_;
-    return *this;
+    init(seed, execution_space().concurrency());
   }
+#endif
 
   void init(uint64_t seed, int num_states) {
     if (seed == 0) seed = uint64_t(1318319);
```

Comment:
IMO this init method should use the seed passed to it as is without doing this check on 0. 
A proper seed should be put inside the constructors directly otherwise we are separating things that are to follow. 

This is what I mean: 
```
#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
  KOKKOS_FUNCTION
  Random_XorShift64_Pool() {
    // ...
    KOKKOS_IF_ON_HOST(init(uint64_t(1318319), execution_space().concurrency());)
  }

  Random_XorShift64_Pool(uint64_t seed) {
    num_states_ = 0;
    init(seed, execution_space().concurrency());
  }
#else
  Random_XorShift64_Pool(uint64_t seed = 0) {
    num_states_ = 0;
    if (seed == 0){ seed = uint64_t(1318319); }
    init(seed, execution_space().concurrency());
  }
#endif

void init(uint64_t seed, ...)
{ 
  // if (seed == 0) seed = uint64_t(1318319); // remove this 
  // I only want to pad ...
  // ...
}
```

Reply:
Outdated since the PR was updated.